### ISSUE TITLE
Churning on downscaled build `Blake2`

### ISSUE THREAD

**Issue by finestructure** [SwiftPackageIndex/SwiftPackageIndex-Server#2029]
This doesn't seem to be related to the new 5.7 downscaling but rather an existing issue. We churn on `Blake2`, which is 100% downscaled (it was probably very slow or broken). It's triggered but not persisted and therefore triggered again:

```
2022-09-15 19:41:45 | [ INFO ] Triggering 18 builds for package name: Blake2, ref: main [component: trigger-builds]
-- | --
2022-09-15 19:41:45 | [ INFO ] too many pending pipelines (89) [component: trigger-builds]
2022-09-15 19:41:46 | [ CRITICAL ] keyNotFound(CodingKeys(stringValue: "web_url", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"web_url\", intValue: nil) (\"web_url\").", underlyingError: nil)) [component: trigger-builds]
2022-09-15 19:42:46 | [ INFO ] Triggering builds (limit: 2) ... [component: trigger-builds]
2022-09-15 19:42:46 | [ INFO ] Finding missing builds for package id: 474CEE01-F3BB-49FD-8DFB-17DAC30D4C14 [component: trigger-builds]
2022-09-15 19:42:46 | [ INFO ] Finding missing builds for package id: B1536570-E72D-46AB-93D3-5301C4ECA712 [component: trigger-builds]
2022-09-15 19:42:46 | [ INFO ] Triggering 16 builds for package name: Blake2, ref: 0.1.2 [component: trigger-builds]
2022-09-15 19:42:46 | [ INFO ] Triggering 18 builds for package name: Blake2, ref: main [component: trigger-builds]
2022-09-15 19:42:46 | [ INFO ] too many pending pipelines (91) [component: trigger-builds]
2022-09-15 19:42:47 | [ CRITICAL ] keyNotFound(CodingKeys(stringValue: "web_url", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"web_url\", intValue: nil) (\"web_url\").", underlyingError: nil)) [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] Triggering builds (limit: 2) ... [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] Finding missing builds for package id: 474CEE01-F3BB-49FD-8DFB-17DAC30D4C14 [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] Finding missing builds for package id: B1536570-E72D-46AB-93D3-5301C4ECA712 [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] Triggering 16 builds for package name: Blake2, ref: 0.1.2 [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] Triggering 18 builds for package name: Blake2, ref: main [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] too many pending pipelines (94) [component: trigger-builds]
2022-09-15 19:43:48 | [ CRITICAL ] keyNotFound(CodingKeys(stringValue: "web_url", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"web_url\", intValue: nil) (\"web_url\").", underlyingError: nil)) [component: trigger-builds]
```

The jobs are immediately aborted:

- https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/3035625932
- https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/3035625936

```
DOWNSCALED
Error: Build downscaling in effect - skipping build
```

This isn't a huge deal but we should make sure we write those build records so they don't come around until they get trimmed.

**Comment by finestructure** [comment]
I've bumped the downscaling factor to 0.2 for now, let's see if that clear it up. There are successful builds for all Swift versions so it doesn't look like this is a package with builds that are timing out.

**Comment by finestructure** [comment]
https://gitlab.com/finestructure/swiftpackageindex-builder/-/jobs/3035665473 macos-xcodebuild/5.4 finished in 1min. I think we can drop this downscaling entirely.

**Comment by finestructure** [comment]
There is something wrong with this package. It's causing lots of stuck jobs that aren't even killed by the `timeout` wrapper:

<img width="312" alt="CleanShot 2022-09-15 at 20 22 38@2x" src="https://user-images.githubusercontent.com/65520/190480875-3f1a382f-b2fd-430a-8ac6-7d486ba1c48e.png">

Going back to 0 downscaling.

**Comment by daveverwer** [comment]
I remember us downscaling it. I wish I remembered why!

**Comment by finestructure** [comment]
Even with Blake2 we're still getting

```
2022-09-15 20:48:18 | [ CRITICAL ] keyNotFound(CodingKeys(stringValue: "web_url", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"web_url\", intValue: nil) (\"web_url\").", underlyingError: nil)) [component: trigger-builds]
```

I'll add some logging so we see what the body is.

**Comment by finestructure** [comment]
Ok, now it's clear what's happening:

```
[ CRITICAL ] decodingFailed("package: https://github.com/ekazaev/ChatLayout.git\nref: 1.2.6\nplatform: macos-spm\nswift version: 5.4\nversion id: D4B7845E-1E57-4410-87B4-17E4D6FD3A3C\nbody: {\"message\":{\"base\":[\"Too many pipelines created in the last minute. Try again later.\"]}}\nerror: keyNotFound(CodingKeys(stringValue: \"web_url\", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: \"No value associated with key CodingKeys(stringValue: \\\"web_url\\\", intValue: nil) (\\\"web_url\\\").\", underlyingError: nil))") [component: trigger-builds]
```

I've not seen that before but it also not something we'd have been able to see with our logging.

**Comment by finestructure** [comment]
Ok, I think the problem with the churning was that Blake2 had many missing builds:

```
2022-09-15 19:43:47 | [ INFO ] Triggering 16 builds for package name: Blake2, ref: 0.1.2 [component: trigger-builds]
2022-09-15 19:43:47 | [ INFO ] Triggering 18 builds for package name: Blake2, ref: main [component: trigger-builds]
```

and triggering 34 when we're already at a high rate causes the `Too many pipelines created in the last minute.` consistently.

The problem is that the error bails out of the whole batch save and doesn't even record back the few that might have made it. So in the end even though we might trigger a few builds, the "stubs" aren't recorded.

This would sort itself out, because even if the stub isn't recorded, the eventual build record would arrive and reduce the trigger count. But because Blake2 is a problematic package, we'd downscaled them - and we don't send build reports for downscaled builds!

The problem seemed to unwedge itself when I lifted the downscaling - the number actually went down a little from 34 missing builds. But then Blake2's build problems struck and took down almost all builders.

I've prepared a change to make sure we save the partial build stubs. This should prevent the churning.
